presto-main/src/main/java/io/prestosql/sql/planner/EffectivePredicateExtractor.java

```diff
@@ -272,6 +281,61 @@ public Expression visitJoin(JoinNode node, Void context)
             }
         }
 
+        @Override
+        public Expression visitValues(ValuesNode node, Void context)
+        {
+            if (node.getOutputSymbols().isEmpty()) {
+                return TRUE_LITERAL;
+            }
+
+            ImmutableMap.Builder<Symbol, Domain> domains = ImmutableMap.builder();
+
+            for (int column = 0; column < node.getOutputSymbols().size(); column++) {
+                Symbol symbol = node.getOutputSymbols().get(column);
+                Type type = types.get(symbol);
+
+                ImmutableList.Builder<Object> builder = ImmutableList.builder();
+                boolean hasNull = false;
+                for (int row = 0; row < node.getRows().size(); row++) {
+                    Expression value = node.getRows().get(row).get(column);
+
+                    if (!DeterminismEvaluator.isDeterministic(value)) {
+                        return TRUE_LITERAL;
+                    }
+
+                    ExpressionInterpreter interpreter = ExpressionInterpreter.expressionOptimizer(value, metadata, session, typeAnalyzer.getTypes(session, types, value));
```

Comment:
maybe `ValuesNode` should only contain `Literal` as rows (I think it implicitly does)

Reply:
Eventuallly, we should probably have a physical vs logical values node or something else to represent that. Currently, a values node inside a correlated subquery could have a reference to the outer scope.

Reply:
> Currently, a values node inside a correlated subquery could have a reference to the outer scope.

That's a problem on it's own. See: https://github.com/prestosql/presto/issues/473

Reply:
It’s not a “problem” per se. It’s legitimate for a query to reference variables from the enclosing scope. After all, filters, projections, subqueries, etc, are nothing but capturing lamba expressions (if you squint a little)

Reply:
> After all, filters, projections, subqueries, etc, are nothing but capturing lamba expressions (if you squint a little)

Still, I don't think correlated symbols should appear out of nothing in any node that contains `Expression`, but should rather be injected into subquery via some kind of special `ProjectCorrelations` node. This way subquery support can be narrowed down in various places.

Reply:
That would be a “bind” operation, just like lambda(x) { ... } binds x.

The next generation of the IR that I’ve been prototyping for a while will likely unify both concepts. Anyway, topic for another time :)